utils/aeromatic++/Systems/Controls.cpp

```diff
@@ -152,75 +152,82 @@ void CableControls::set(const float* cg_loc)
-
-    float Re = 150000;
-    float Cf = 1.328f/sqrtf(Re);
-
+// https://www.fzt.haw-hamburg.de/pers/Scholz/HOOU/AircraftDesign_13_Drag.pdf
     float M2, M = 0.01f;
-    float fM = 1.0f - 0.08f*powf(M, 1.45f);
 
-    // Sfus: Fuselage wetted area
-    float D = _aircraft->get_fuselage_diameter();
-    float L = _aircraft->_length;
-    float fLD = 1.0f + 60.0f/powf(L/D, 3.0f) + 0.0025f*(L/D);
-    float Sfus = 0.75f*PI*D*L;
-    float CD0f = Cf*fLD*fM*(Sfus/Sw);
-
-    // Main Wing
-    Cf = 0.2f*Cf + 0.8f*0.455f/powf(logf(Re),2.58f);
-    float CDminw = 0.004f;
-    float fD = powf(CDminw/0.004f, 0.4f);
-    float fTC = 1.0f + 2.7f*TC + 100.0f*powf(TC, 4.0f);
-    float Swing = 2.0f*(1.0f + 0.5f*TC)*bw*cbarw;
-    float CD0w = Cf*fM*fTC*fD*(Swing/Sw);
-    CD0w *= (1.0f - sinf(sweep));
-
-    // Horizontal tail
-    float cbarh = _aircraft->_htail.chord_mean;
-    float TCh = _aircraft->_htail.thickness/cbarh;
-    float CDminh = 0.002f;
-    float fDh = powf(CDminh/0.004f, 0.4f);
-    float fTCh = 1.0f + 2.7f*TCh + 100.0f*powf(TCh, 4.0f);
-    float CD0h = Cf*fM*fTCh*fDh*(Swing/Sw);
-    CD0h *= (1.0f - sinf(_aircraft->_htail.sweep));
-
-    // Vertical tail
-    float bv = _aircraft->_vtail.span;
-    float cbarv = _aircraft->_vtail.chord_mean;
-    float TCv = _aircraft->_vtail.thickness/cbarv;
-    float CDminv = 0.002f;
-    float fDv = powf(CDminv/0.004f, 0.4f);
-    float fTCv = 1.0f + 2.7f*TCv + 100.0f*powf(TCv, 4.0f);
-    float Svtail = 2.0f*(1.0f + 0.5f*TCv)*bv*cbarv;
-    float CD0v = Cf*fM*fTCv*fDv*(Svtail/Sv);
-    CD0v *= (1.0f - sinf(_aircraft->_vtail.sweep));
-
-    float Kc = 1.05f;
+    float Cf;
     switch(_aircraft->_atype)
     {
     case LIGHT:
-        if (_aircraft->_no_engines > 0) Kc = 1.3f;
+        if (_aircraft->_no_engines == 0) Cf = 0.0030f;
+        else if (_aircraft->_no_engines == 1) Cf = 0.0055f;
+        else Cf = 0.0045f;
         break;
-    case PERFORMANCE:
     case PROP_TRANSPORT:
-        Kc = 1.2f;
+        Cf = 0.0035f;
         break;
-    case FIGHTER:
     case JET_TRANSPORT:
-        Kc = 1.1f;
+        Cf = 0.0030f;
         break;
-//  case BIPLANE:
-    default:
-        Kc = 1.5f;
+    case PERFORMANCE:
+        Cf = 0.0035f;
+        break;
+    case FIGHTER:
+        Cf = 0.0030f;
         break;
-    }
 #if 0
- printf("CD0: %f (%f), Kc: %f, CD0f: %f, CD0w: %f, CD0h: %f, CD0v: %f\n", _aircraft->_CD0, Kc*(CD0f + CD0w + CD0h + CD0v), Kc, CD0f, CD0w, CD0h, CD0v);
+    case BIPLANE:
+        Cf = 0.0065f;
+        break;
 #endif
-    _aircraft->_CD0 = Kc*(CD0f + CD0w + CD0h + CD0v);
+    default:
+        Cf = 0.0030f;
+        break;
+    }
+
+    // Fuselage
+    float Df = _aircraft->get_fuselage_diameter();
+    float Lf = _aircraft->_length;
+    float fr = Lf/Df; // fuselage finess ratio
+
+    float Qf = 1.0f;
+    float FFf = 1.0f + 60.0f/powf(fr, 3.0f) + fr/400.0f;
+    float Swet_f = PI*Df*Lf*pow(1.0-2.0/fr, 0.667f)*(1.0 + 1.0/(fr*fr));
```

Comment:
GitHub CodeQL raises an error on that line because `PI` is (I think) a `double` while the rest of the expression is made of `float` variable. Would you mind convert `PI` to a float in this expression ?
```suggestion
    float Swet_f = ((float)PI)*Df*Lf*pow(1.0-2.0/fr, 0.667f)*(1.0 + 1.0/(fr*fr));
```

Reply:
Odd, PI is defined float explicitly:
types.h:#define PI			3.14159265358979323846f

Maybe it's too exact for a float and turned to a double automatically?

Reply:
May be or may be it's the `pow()` function that needs to be changed to `powf()` to return a `float` as well as the constants that should be appended with a `f` ?

Reply:
Shoot, I noticed forgot about the pow function. Hold on.